### PR TITLE
Feature/4798 custom confirmation modal

### LINKED ISSUE #4798: Using custom confirmation modals
### Thema / Theme

Frontend

### Omschrijving / Description

Replacing the `window.confirm(...)` confirmations with a custom confirmation that uses our Modal component

### Added value / Toegevoegde waarde

_No response_

### Aanvullende opmerkingen / Additional context

_No response_

### PR DESCRIPTION
**Pull request by robinmolen**
Closes #4798

**Changes**

Replacing all `window.confirm(...)` confirmations with a new custom React hook (`useConfirm`). This new hook is using our Modal component to display the confirmation message. These changes happend to keep the application UI uniform.

**Checklist**

Check off the items that are completed or not relevant.

- Impact on features

  - [X] Checked copying a form
  - [X] Checked import/export of a form
  - [X] Config checks in the configuration overview admin page
  - [X] Problem detection in the admin email digest is handled

- Release management

  - [X] I have labelled the PR as "needs-backport" accordingly

- I have updated the translations assets (you do NOT need to provide translations)

  - [X] Ran `./bin/makemessages_js.sh`
  - [X] Ran `./bin/compilemessages_js.sh`

- Commit hygiene

  - [X] Commit messages refer to the relevant Github issue
  - [X] Commit messages explain the "why" of change, not the how

### CODE REVIEW

**Comment by stevenbal** [comment]
@robinmolen can you fix the merge conflicts? Also not sure if the failing tests are related to the code changes?

**Comment by robinmolen** [comment]
There is a new window.confirm added in a recent commit, in the `/form_design/form-creation-form.js`. I'll replace that one now with the new `useConfirm`

**Comment by robinmolen** [comment]
I've replaced the `.../form_design/form-creation-form.js` `window.confirm` with the `useConfirm`. With that, all window.confirm's have been replaced :)

File: src/openforms/js/components/admin/form_design/useConfirm.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      defaultMessage: 'Cancel',
+    });
+    return (
+      <Modal
+        title={title}
+        isOpen={promise !== null}
+        contentModifiers={['with-form', 'small']}
+        closeModal={handleCancel}
+      >
+        <div style={{flexGrow: 1}}>{message}</div>
```
Comment (stevenbal) [inline src/openforms/js/components/admin/form_design/useConfirm.js:46]: Maybe a class for this should be added where this styling is set instead?

File: src/openforms/js/components/admin/form_design/registrations/zgw/fields/ZGWAPIGroup.js [hunk 2]
```diff
@@ -44,13 +44,16 @@ const ZGWAPIGroup = ({apiGroupChoices, onChangeCheck}) => {
         }
         noManageChildProps
       >
-        <ReactSelect
+        <SelectWithoutFormik
```
Comment (stevenbal) [inline src/openforms/js/components/admin/form_design/registrations/zgw/fields/ZGWAPIGroup.js:47]: Are these changes intentional? I'm not sure if we want to remove formik for this field

File: src/openforms/js/components/admin/form_design/useConfirm.stories.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      ) : null}
+      <ConfirmationModal />
+    </div>
+  );
+};
+
+export default {
+  title: 'Admin / Custom / UseConfirm',
+  render: () => <ButtonWithUseConfirm />,
+  component: useConfirm,
```
Comment (stevenbal) [inline src/openforms/js/components/admin/form_design/useConfirm.stories.js:32]: It might be good to add a basic test here to open the modal and check that it has the expected text

File: src/openforms/js/components/admin/form_design/useConfirm.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        <div className="button-group">
```
Comment (stevenbal) [inline src/openforms/js/components/admin/form_design/useConfirm.js:47]: We could maybe reuse `submit-row`, to remove the need for extra styling for button-group

image

**Review by stevenbal — CHANGES_REQUESTED** [review]

File: src/openforms/js/components/admin/form_design/useConfirm.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    promise?.resolve(true);
+    handleClose();
+  };
+
+  const handleCancel = () => {
+    promise?.resolve(false);
+    handleClose();
+  };
+
+  const ConfirmationModal = () => {
```
[the remark this answers is not in the document]
Comment (robinmolen) [reply src/openforms/js/components/admin/form_design/useConfirm.js:29]: Yeah this is the normal `small` modal modifier with a height of 25vh, but for the confirmations this is indeed a bit much..
I'll a `dialog` modifier option to the modal component with a bit nicer styling

File: src/openforms/js/components/admin/form_design/registrations/zgw/fields/ZGWAPIGroup.js (hunk 2 above)
Comment (robinmolen) [reply src/openforms/js/components/admin/form_design/registrations/zgw/fields/ZGWAPIGroup.js:47]: Yes this is intentional. The ReactSelect should set the value after the confirmation is accepted. The `SelectWithFormik` (the component that is exported as the default, aka `ReactSelect`) doesn't allow custom onChange handling, seeing how Formik should control this component.

So another solution for this would be to add a `onChange` handler to the `SelectWithFormik`, but that would (in my opinion) kinda defeat the purpose of the component..

This field still uses Formik, but the value/change handling is now done by the component, instead of the Select

File: src/openforms/js/components/admin/form_design/useConfirm.stories.js (hunk 3 above)
Comment (robinmolen) [reply src/openforms/js/components/admin/form_design/useConfirm.stories.js:32]: Ah yeah check, will do!

File: src/openforms/js/components/admin/form_design/useConfirm.js (hunk 4 above)
Comment (robinmolen) [reply src/openforms/js/components/admin/form_design/useConfirm.js:47]: Yeah i don't really like the styling of the `submit-row`. The weird extra border and background, extra padding and the uppercase button text.

I get your point, and maybe a modal action bar would be nice (in my head more like a bootstrap modal, where the padding between the `submit-row` and the modal would be removed). But then i would like to change that for all the modals.

For now i think the button group looks quite a bit better

File: src/openforms/js/components/admin/form_design/useConfirm.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (robinmolen) [reply src/openforms/js/components/admin/form_design/useConfirm.js:29]: I've made it a bit differently. The modal component now has a new `confirmation` modifier. This applies some styling changes to make the modal "grow" with the content (i.e. no fixed height). This also allows the inline flexGrow styling to be removed

File: src/openforms/js/components/admin/form_design/useConfirm.js (hunk 1 above)
Comment (robinmolen) [reply src/openforms/js/components/admin/form_design/useConfirm.js]: With some modal styling changes is this inline styling no longer needed 👍 

Extra info: https://github.com/open-formulieren/open-forms/pull/4814#discussion_r1838513591

File: src/openforms/js/components/admin/form_design/useConfirm.js (hunk 4 above)
Comment (robinmolen) [reply src/openforms/js/components/admin/form_design/useConfirm.js]: I've removed the `button-group` styling and added some modal specific styling. So this still doesn't use the `submit-row` class

File: src/openforms/js/components/admin/form_design/useConfirm.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      >
+        <p>{message}</p>
+        <div className="react-modal__actions">
+          <ActionButton text={confirmBtnText} className="default" onClick={handleConfirm} />
+          <ActionButton text={cancelBtnText} onClick={handleCancel} />
+        </div>
+      </Modal>
+    );
+  };
+  return [ConfirmationModal, confirm];
```
Comment (sergei-maertens) [inline src/openforms/js/components/admin/form_design/useConfirm.js:54]: I remember reading that dynamically defining React components is an anti-pattern, and have tried this in the past in other places too and the conclusion was that it changes all the time/re-renders because there's no stable identity.

Utlimately, the proper solution is to return props to use for the component, and not return the whole component as is.

I'm afraid this approach here suffers from those same issues, and it may have been one of the reasons you were ending up in infinite render loops :thinking:

Comment (robinmolen) [reply src/openforms/js/components/admin/form_design/useConfirm.js:54]: Ah okay, that sound very interesting... Yeah then this very well might have played a bigger part in the whole infinite render issue..
I'll look into the article, and change the hook

**Review by sergei-maertens — COMMENTED** [review]
Holding of *a little* longer because of the SDK context/decorator for storybook I mentioned on Slack. I think it would clean up the new/updated interaction tests, and I'm curious if that formik onChange stuff is still necessary, since what it currently does is stop tracking the `isTouched` state and once we refactor to not roll our validation errors/mechanics 100% on our own, that will definitely cause some problems then :)

File: src/openforms/js/components/admin/form_design/registrations/zgw/fields/ZGWAPIGroup.js (hunk 2 above)
Comment (robinmolen) [reply src/openforms/js/components/admin/form_design/registrations/zgw/fields/ZGWAPIGroup.js:47]: Yeah, its still the case that the `SelectWithFormik` (at the moment) doesn't allow custom onChange handling. This is needed, to trigger the confirmation modal onChange, and to save the data to formik after the confirmation is accepted.

Comment (robinmolen) [reply src/openforms/js/components/admin/form_design/registrations/zgw/fields/ZGWAPIGroup.js:47]: Nevermind... I just noticed (with a comment from Sergei) that you can overwrite the props of the `SelectWithFormik`..
So yeah, this isn't needed and i'll revert this

**Review by sergei-maertens — APPROVED** [review]
bellissimo
